# Capstone Project Notebook

# Part I

## 1. Use pandas, or the BeautifulSoup package, or any other way you are comfortable with to transform the data in the table on the Wikipedia page into pandas dataframe.

### Using Xpath and Pandas to transform the data in the table on the Wikipedia page into pandas dataframe

### Import Libraries

<p>We will need requests for getting the HTML contents of the website and lxml.html for parsing the relevant fields. Finally, we will store the data on a Pandas Dataframe.</p>

In [1]:
import pandas as pd
import requests
import lxml.html as lh

In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
#Create a handle, page, to handle the contents of the website
page = requests.get(url)
#Store the contents of the website under doc
doc = lh.fromstring(page.content)
#Parse data that are stored between <tr>..</tr> of HTML
tr_elements = doc.xpath('//tr')


<p>For sanity check, ensure that all the rows have the same width. If not, we probably got something more than just the table.</p>

In [3]:
#Check the length of the first 12 rows
[len(T) for T in tr_elements[:12]]

[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]

Looks like all our rows have exactly 3 columns. This means all the data collected on tr_elements are from the table.

### Parse Table Header

Next, let’s parse the first row as our header.

In [4]:
#col.append((name,[]))
    
    #print("Total score for %s is %s" % (name, score))
#tr_elements = doc.xpath('//tr')
#Create empty list
col=[]
i=0
#For each row, store each first element (header) and an empty list
for t in tr_elements[0]:
    i+=1
    name=t.text_content()
    print ("%d : %s" % (i,name))
    col.append((name,[]))      
    
           

1 : Postcode
2 : Borough
3 : Neighbourhood



### Creating Pandas DataFrame

Each header is appended to a tuple along with an empty list.

In [5]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    #T=tr_elements[j]
    f=0
    for field in tr_elements[j]:
        f+=1
        name=field.text_content()
        if name=='Not assigned':
            break
        print(name)

M1A
M2A
M3A
North York
Parkwoods

M4A
North York
Victoria Village

M5A
Downtown Toronto
Harbourfront

M6A
North York
Lawrence Heights

M6A
North York
Lawrence Manor

M7A
Downtown Toronto
Queen's Park

M8A
M9A
Etobicoke
Islington Avenue

M1B
Scarborough
Rouge

M1B
Scarborough
Malvern

M2B
M3B
North York
Don Mills North

M4B
East York
Woodbine Gardens

M4B
East York
Parkview Hill

M5B
Downtown Toronto
Ryerson

M5B
Downtown Toronto
Garden District

M6B
North York
Glencairn

M7B
M8B
M9B
Etobicoke
Cloverdale

M9B
Etobicoke
Islington

M9B
Etobicoke
Martin Grove

M9B
Etobicoke
Princess Gardens

M9B
Etobicoke
West Deane Park

M1C
Scarborough
Highland Creek

M1C
Scarborough
Rouge Hill

M1C
Scarborough
Port Union

M2C
M3C
North York
Flemingdon Park

M3C
North York
Don Mills South

M4C
East York
Woodbine Heights

M5C
Downtown Toronto
St. James Town

M6C
York
Humewood-Cedarvale

M7C
M8C
M9C
Etobicoke
Bloordale Gardens

M9C
Etobicoke
Eringate

M9C
Etobicoke
Markland Wood

M9C
Etobicoke
Old Burnhamt

In [6]:
#Since out first row is the header, data is stored on the second row onwards
for j in range(1,len(tr_elements)):
    #T is our j'th row
    T=tr_elements[j]
    
    #If row is not of size 3, the //tr data is not from our table 
    if len(T)!=3:
        break
    
    #i is the index of our column
    i=0
    
    #Iterate through each element of the row
    for t in T.iterchildren():
        
        data=t.text_content() 
        #Check if row is empty
        if i>0:
            #Convert any numerical value to integers
            try:
                data=int(data)
            except:
                pass
        #Append the data to the empty list of the i'th column
        col[i][1].append(data)
        #Increment i for the next column
        i+=1

Just to be sure, let’s check the length of each column. Ideally, they should all be the same.

In [7]:
[len(C) for (title,C) in col]

[287, 287, 287]

Perfect! This shows that each of our 3 columns has exactly 287 values.

Now we are ready to create the DataFrame:

In [8]:
Dict={title:column for (title,column) in col}
df=pd.DataFrame(Dict)

Looking at the top 5 cells on the DataFrame:

In [9]:
df.head(5)

Postcode           Borough     Neighbourhood\n
0      M1A      Not assigned      Not assigned\n
1      M2A      Not assigned      Not assigned\n
2      M3A        North York         Parkwoods\n
3      M4A        North York  Victoria Village\n
4      M5A  Downtown Toronto      Harbourfront\n

## 2. The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood

### Renaming column names 

In [10]:
df.rename(columns={'Postcode': 'PostalCode', 'Borough': 'Borough', 'Neighbourhood\n': 'Neighborhood'}, inplace=True)
df.head()

PostalCode           Borough        Neighborhood
0        M1A      Not assigned      Not assigned\n
1        M2A      Not assigned      Not assigned\n
2        M3A        North York         Parkwoods\n
3        M4A        North York  Victoria Village\n
4        M5A  Downtown Toronto      Harbourfront\n

Strip Trailing Space of the Neighbourhood column

In [11]:
df['Neighborhood'] = df['Neighborhood'].str.rstrip()
df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

In [12]:
df.dtypes

PostalCode      object
Borough         object
Neighborhood    object
dtype: object

In [13]:
df.shape

(287, 3)

## 3.Only processing the cells that have an assigned borough. Ignoring cells with a borough that is Not assigned.

In [14]:
#df = df[df['Borough'] != 'Not assigned'] 
# Get names of indexes for which column Borough has value "Not assigned"
indexNames = df[df['Borough'] =='Not assigned'].index

# Delete these row indexes from dataFrame
df.drop(indexNames , inplace=True)        
df.head()

PostalCode           Borough      Neighborhood
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront
5        M6A        North York  Lawrence Heights
6        M6A        North York    Lawrence Manor

In [15]:
df.shape

(210, 3)

## 4. If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

In [16]:
#for index_label, row_series in df.iterrows():
    #if df.at[index_label , 'Neighborhood'] == 'Not assigned':
        #df.at[index_label , 'Neighborhood'] = df.at[index_label , 'Borough']
        
df.loc[df['Neighborhood'] =='Not assigned' , 'Neighborhood'] = df['Borough']
df.head(10)
 

PostalCode           Borough      Neighborhood
2         M3A        North York         Parkwoods
3         M4A        North York  Victoria Village
4         M5A  Downtown Toronto      Harbourfront
5         M6A        North York  Lawrence Heights
6         M6A        North York    Lawrence Manor
7         M7A  Downtown Toronto      Queen's Park
9         M9A         Etobicoke  Islington Avenue
10        M1B       Scarborough             Rouge
11        M1B       Scarborough           Malvern
13        M3B        North York   Don Mills North

In [17]:
df.shape

(210, 3)

## 5.Rows with same postalcode will combined into one row with the neighborhoods separated with a comma

In [18]:

#df['Neighborhood'] = df.groupby(['PostalCode'])['Neighborhood'].transform(lambda x: ','.join(x))
#df.drop_duplicates(inplace=True)

result = df.groupby(['PostalCode', 'Borough'], sort=False).agg(', '.join)
df_new = result.reset_index()
df_new.head(10)

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto                      Harbourfront
3        M6A        North York  Lawrence Heights, Lawrence Manor
4        M7A  Downtown Toronto                      Queen's Park
5        M9A         Etobicoke                  Islington Avenue
6        M1B       Scarborough                    Rouge, Malvern
7        M3B        North York                   Don Mills North
8        M4B         East York   Woodbine Gardens, Parkview Hill
9        M5B  Downtown Toronto          Ryerson, Garden District

## 6.Use the .shape method to print the number of rows of your dataframe.

In [19]:
df_new.shape

(103, 3)

# Part 2

### Using csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

In [20]:
!wget -q -O 'Toronto_long_lat_data.csv'  http://cocl.us/Geospatial_data

In [21]:
df_lon_lat = pd.read_csv('Toronto_long_lat_data.csv')
df_lon_lat.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [22]:
df_lon_lat.columns = ['PostalCode', 'Latitude', 'Longitude']
df_lon_lat

PostalCode   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
5          M1J  43.744734 -79.239476
6          M1K  43.727929 -79.262029
7          M1L  43.711112 -79.284577
8          M1M  43.716316 -79.239476
9          M1N  43.692657 -79.264848
10         M1P  43.757410 -79.273304
11         M1R  43.750072 -79.295849
12         M1S  43.794200 -79.262029
13         M1T  43.781638 -79.304302
14         M1V  43.815252 -79.284577
15         M1W  43.799525 -79.318389
16         M1X  43.836125 -79.205636
17         M2H  43.803762 -79.363452
18         M2J  43.778517 -79.346556
19         M2K  43.786947 -79.385975
20         M2L  43.757490 -79.374714
21         M2M  43.789053 -79.408493
22         M2N  43.770120 -79.408493
23         M2P  43.752758 -79.400049
24         M2R  43.782736 -79.442259
25         M3A  43.753259 -79.329656
26         M3B  43.745906 -79.352188
27         M3C  43.725900 -79.340923
28         M3H  43.754328 -79.442259
29         M3J  43.767980 -79.487262
..         ...        ...        ...
73         M6C  43.693781 -79.428191
74         M6E  43.689026 -79.453512
75         M6G  43.669542 -79.422564
76         M6H  43.669005 -79.442259
77         M6J  43.647927 -79.419750
78         M6K  43.636847 -79.428191
79         M6L  43.713756 -79.490074
80         M6M  43.691116 -79.476013
81         M6N  43.673185 -79.487262
82         M6P  43.661608 -79.464763
83         M6R  43.648960 -79.456325
84         M6S  43.651571 -79.484450
85         M7A  43.662301 -79.389494
86         M7R  43.636966 -79.615819
87         M7Y  43.662744 -79.321558
88         M8V  43.605647 -79.501321
89         M8W  43.602414 -79.543484
90         M8X  43.653654 -79.506944
91         M8Y  43.636258 -79.498509
92         M8Z  43.628841 -79.520999
93         M9A  43.667856 -79.532242
94         M9B  43.650943 -79.554724
95         M9C  43.643515 -79.577201
96         M9L  43.756303 -79.565963
97         M9M  43.724766 -79.532242
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242
100        M9R  43.688905 -79.554724
101        M9V  43.739416 -79.588437
102        M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [23]:
df_lon_lat.shape

(103, 3)

In [24]:
dfinal = df_new.merge(df_lon_lat, on="PostalCode", how = 'inner')
dfinal

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
..         ...               ...   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                                         Harbourfront  43.654260 -79.360636  
3                     Lawrence Heights, Lawrence Manor  43.718518 -79.464763  
4                                         Queen's Park  43.662301 -79.389494  
5                                     Islington Avenue  43.667856 -79.532242  
6                                       Rouge, Malvern  43.806686 -79.194353  
7                                      Don Mills North  43.745906 -79.352188  
8                      Woodbine Gardens, Parkview Hill  43.706397 -79.309937  
9                             Ryerson, Garden District  43.657162 -79.378937  
10                                           Glencairn  43.709577 -79.445073  
11   Cloverdale, Islington, Martin Grove, Princess ...  43.650943 -79.554724  
12              Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
13                    Flemingdon Park, Don Mills South  43.725900 -79.340923  
14                                    Woodbine Heights  43.695344 -79.318389  
15                                      St. James Town  43.651494 -79.375418  
16                                  Humewood-Cedarvale  43.693781 -79.428191  
17   Bloordale Gardens, Eringate, Markland Wood, Ol...  43.643515 -79.577201  
18                   Guildwood, Morningside, West Hill  43.763573 -79.188711  
19                                         The Beaches  43.676357 -79.293031  
20                                         Berczy Park  43.644771 -79.373306  
21                               